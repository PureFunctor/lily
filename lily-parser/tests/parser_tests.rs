use lily_parser::parse_top_level;

#[test]
pub fn top_level_0() {
    let source = r"
infixl 4 add as +

example = a + b + c
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_1() {
    let source = r"
example = a + b + c
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_2() {
    let source = r"
infixl 1 add as +
infixl 2 add as *

example = a + b * (c + d) + e
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_3() {
    let source = r"
infixl 1 add as +

example = if if a then b else c then if d then e else f else if g then h else i
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_4() {
    let source = r"
example = f a b
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_5() {
    let source = r"
example = f if a then b else c d

example = f (if a then b else c) d
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}
