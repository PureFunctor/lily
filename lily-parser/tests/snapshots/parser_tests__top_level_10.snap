---
source: lily-parser/tests/parser_tests.rs
expression: parse_top_level(source)
---
Ok(
    Module {
        declarations: [
            Declaration {
                begin: 31,
                end: 59,
                kind: TypeDeclaration(
                    "example",
                    Ty {
                        begin: 41,
                        end: 59,
                        kind: BinaryOperator(
                            Ty {
                                begin: 41,
                                end: 42,
                                kind: Application(
                                    Ty {
                                        begin: 41,
                                        end: 42,
                                        kind: Constructor(
                                            "A",
                                        ),
                                    },
                                    [
                                        Ty {
                                            begin: 43,
                                            end: 44,
                                            kind: Variable(
                                                "a",
                                            ),
                                        },
                                    ],
                                ),
                            },
                            "->",
                            Ty {
                                begin: 48,
                                end: 59,
                                kind: BinaryOperator(
                                    Ty {
                                        begin: 48,
                                        end: 49,
                                        kind: Constructor(
                                            "B",
                                        ),
                                    },
                                    "->",
                                    Ty {
                                        begin: 53,
                                        end: 59,
                                        kind: BinaryOperator(
                                            Ty {
                                                begin: 53,
                                                end: 54,
                                                kind: Constructor(
                                                    "C",
                                                ),
                                            },
                                            "->",
                                            Ty {
                                                begin: 58,
                                                end: 59,
                                                kind: Application(
                                                    Ty {
                                                        begin: 58,
                                                        end: 59,
                                                        kind: Constructor(
                                                            "D",
                                                        ),
                                                    },
                                                    [
                                                        Ty {
                                                            begin: 60,
                                                            end: 65,
                                                            kind: Parenthesized(
                                                                Ty {
                                                                    begin: 61,
                                                                    end: 62,
                                                                    kind: Application(
                                                                        Ty {
                                                                            begin: 61,
                                                                            end: 62,
                                                                            kind: Variable(
                                                                                "d",
                                                                            ),
                                                                        },
                                                                        [
                                                                            Ty {
                                                                                begin: 63,
                                                                                end: 64,
                                                                                kind: Variable(
                                                                                    "e",
                                                                                ),
                                                                            },
                                                                        ],
                                                                    ),
                                                                },
                                                            ),
                                                        },
                                                    ],
                                                ),
                                            },
                                        ),
                                    },
                                ),
                            },
                        ),
                    },
                ),
            },
        ],
    },
)
